### PR TITLE
dev/core#1945 Fix recur access regression

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Fix regression on accessing recurring contributions

Before
----------------------------------------
Per https://lab.civicrm.org/dev/core/-/issues/1945

After
----------------------------------------
Fixed

Technical Details
----------------------------------------
I looked at an extraction in getContactID but just couldn't make anything that was clean

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Core/Form.php [hunk 1]
```diff
@@ -2656,4 +2656,19 @@ public function setSelectedChild($default = NULL) {
...
+   * @return string|bool
+   * @throws \CRM_Core_Exception
+   */
+  protected function getContactFromUrl(): bool {
+    $contactID = CRM_Utils_Request::retrieve('cid', 'Positive');
+    if (!$contactID) {
+      return FALSE;
+    }
+    $userChecksum = CRM_Utils_Request::retrieve('cs', 'String');
+    return (int) CRM_Contact_BAO_Contact_Utils::validChecksum($contactID, $userChecksum) ? $userChecksum : $contactID;
```
Comment (seamuslee001) [inline CRM/Core/Form.php:2671]: So if the checksum doesn't validate we're still returning the contact id?

Comment (eileenmcnaughton) [reply CRM/Core/Form.php:2671]: Hmm - that might not be right - let me look again

Comment (eileenmcnaughton) [reply CRM/Core/Form.php]: @seamuslee001 I think this is better

**Comment by seamuslee001** [comment]
This looks correct to me @eileenmcnaughton do you think I should wait for the reporter to get back to us or just merge it in?

**Comment by eileenmcnaughton** [comment]
@seamuslee001 it would be good to merge - I just checked the typing. Probably an r-run to check checksum still works would make sense

File: CRM/Contribute/Form/ContributionRecur.php [hunk 2]
```diff
@@ -206,7 +206,7 @@ protected function isSelfService() {
     }
     $this->selfService = FALSE;
     if (!CRM_Core_Permission::check('edit contributions')) {
-      if ($this->_subscriptionDetails->contact_id != $this->getContactID()) {
+      if ((int) $this->_subscriptionDetails->contact_id !== $this->getContactFromUrl()()) {
```
Comment (seamuslee001) [inline CRM/Contribute/Form/ContributionRecur.php:209]: @eileenmcnaughton I think there is an extra () here

**Comment by seamuslee001** [comment]
@eileenmcnaughton I tested this but there is a bug in this apart from the issue I have raised a sample URL for the updating of subscription is 

http://47-test/civicrm/contribute/updaterecur?reset=1&crid=2&cs=4a9e71f0006bda355d749f0826e404ca_1597720187_inf

note how it doesn't have the cid in the url rather the crid. I note that the former code used $this->_subscriptionDetails->contact_id  as part of the check in validChecksum

**Comment by seamuslee001** [comment]
The url I generated used the params https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/BAO/ContributionPage.php#L586 which were passed through to https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/Payment.php#L1693

File: CRM/Core/Form.php [hunk 3]
```diff
@@ -2656,4 +2659,26 @@ public function setSelectedChild($default = NULL) {
...
+  protected function getContactIDIfAccessingOwnRecord() {
+    $contactID = (int) CRM_Utils_Request::retrieve('cid', 'Positive', $this);
+    if (!$contactID) {
+      return FALSE;
+    }
+    if ($contactID === $this->getLoggedInUserContactID()) {
+      return $contactID;
+    }
+    $userChecksum = CRM_Utils_Request::retrieve('cs', 'String');
+    return CRM_Contact_BAO_Contact_Utils::validChecksum($contactID, $userChecksum) ? $contactID : FALSE;
```
Comment (seamuslee001) [inline CRM/Core/Form.php:2681]: @eileenmcnaughton this still looks like it will fail because $contactID will be empty for anon user

Comment (eileenmcnaughton) [reply CRM/Core/Form.php:2681]: @seamuslee001 cid is set in preProcess rather than the url

**Comment by seamuslee001** [comment]
I have tested this and confirmed it works as expected now

**Pull request merged**